src/mame/sony_news/news_r4k.cpp

```diff
@@ -0,0 +1,1261 @@
+}
+
+/*
+ * ram_r
+ *
+ * Method that returns the byte at `offset` in main memory.
+ * TODO: Memory controller logic for non-64MB configurations
+ */
+uint8_t news_r4k_state::ram_r(offs_t offset)
+{
+	uint8_t result = 0xff;
+	if ((offset <= 0x1ffffff) || (m_ram_map_shift && offset <= 0x3ffffff))
+	{
+		result = m_ram->read(offset);
+	}
+	else if (!m_ram_map_shift && offset >= 0x7f00000) // upper 32MB before it is mapped into contiguous space
+	{
+		result = m_ram->read(offset - 0x5f00000);
+	}
+	else
+	{
+		LOGMASKED(LOG_GENERAL, "Unmapped RAM read attempted at offset 0x%x\n", offset);
+	}
+	return result;
+}
+
+/*
+ * ram_w
+ *
+ * Method that writes the byte `data` at `offset` in main memory.
+ * TODO: Memory controller logic for non-64MB configurations
+ */
+void news_r4k_state::ram_w(offs_t offset, uint8_t data)
+{
+	if ((offset <= 0x1ffffff) || (m_ram_map_shift && offset <= 0x3ffffff))
+	{
+		m_ram->write(offset, data);
+	}
+	else if (!m_ram_map_shift && offset >= 0x7f00000) // upper 32MB before it is mapped into contiguous space
+	{
+		m_ram->write(offset - 0x5f00000, data);
+	}
+	else
+	{
+		LOGMASKED(LOG_GENERAL, "Unmapped RAM write attempted at offset 0x%x (data: 0x%x)\n", offset, data);
+	}
+}
+
+/*
+ * machine_start
+ *
+ * Initialization method called when the machine first starts.
+ */
+void news_r4k_state::machine_start()
+{
+	// Init front panel LEDs
+	m_led.resolve();
+
+	// Save state support
+	save_item(NAME(m_inten));
+	save_item(NAME(m_intst));
+	save_item(NAME(m_int_state));
+	save_item(NAME(m_freerun_timer_val));
+
+	// Allocate hardware timers
+	m_freerun_timer = machine().scheduler().timer_alloc(timer_expired_delegate(FUNC(news_r4k_state::freerun_clock), this));
+	m_timer0_timer = machine().scheduler().timer_alloc(timer_expired_delegate(FUNC(news_r4k_state::timer0_clock), this));
+
+	// Cache for main bus access (APbus address translation)
+	m_cpu->space(0).cache(m_main_cache);
+}
+
+/*
+ * freerun_clock
+ *
+ * Method to handle a tick of the freerunning clock.
+ */
+TIMER_CALLBACK_MEMBER(news_r4k_state::freerun_clock)
+{
+	m_freerun_timer_val++;
+}
+
+/*
+ * timer0_clock
+ *
+ * Method to handle a tick of the TIMER0 hardware clock.
+ */
+TIMER_CALLBACK_MEMBER(news_r4k_state::timer0_clock)
+{
+	// Whenever this timer elapses, we need to trigger the timer0 interrupt
+	m_intst[2] |= irq2_number::TIMER0;
+	int_check();
+}
+
+/*
+ * machine_reset
+ *
+ * Method called whenever the machine is reset.
+ */
+void news_r4k_state::machine_reset()
+{
+	m_freerun_timer_val = 0;
+	m_freerun_timer->adjust(attotime::zero, 0, attotime::from_hz(FREERUN_FREQUENCY));
+	m_timer0_timer->adjust(attotime::zero, 0, attotime::from_hz(TIMER0_FREQUENCY));
+}
+
+/*
+ * init_common
+ *
+ * Initialization method called when the machine first starts.
+ */
+void news_r4k_state::init_common() {}
+
+/*
+ * init_nws5000x
+ *
+ * Initialization method specific to the NWS-5000X emulation.
+ */
+void news_r4k_state::init_nws5000x() { init_common(); }
```

Comment:
Are these really necessary?  You can use `empty_init` if you don’t need to do ROM unscrambling or anything.